File: CPP/No565.cc
/**
 * Created by Xiaozhong on 2020/10/20.
 * Copyright (c) 2020/10/20 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>
#include <unordered_set>
#include <list>

using namespace std;

class Solution {
public:
    int arrayNesting(vector<int> &nums) {
        int ans = 0;
        vector<bool> visited(nums.size(), false);
        for (int i = 0; i < nums.size(); ++i) {
            int path = 0;
            while (!visited[i]) {
                visited[i] = true;
                ++path;
                i = nums[i];
            }
            ans = max(ans, path);
        }
        return ans;
    }
};

int main() {
    Solution s;
    vector<int> nums = {5, 4, 0, 3, 1, 6, 2};
    cout << s.arrayNesting(nums) << endl;
}

File: Java/src/No565.java
/**
 * Java
 *
 * @Description: TODO
 * @Author: cherry
 * @Create on: 2022/7/17
 **/

public class No565 {
    public int arrayNesting(int[] nums) {
        int ans = 0;
        int n = nums.length;
        boolean[] visited = new boolean[n];
        for (int idx = 0; idx < n; ++idx) {
            if (visited[idx]) continue;
            int curr = nums[idx];
            int len = 0;
            while (!visited[curr]) {
                len++;
                visited[curr] = true;
                curr = nums[curr];
            }
            ans = Math.max(ans, len);
        }
        return ans;
    }

    public static void main(String[] args) {
        int[] nums = {5, 4, 0, 3, 1, 6, 2};
        System.out.println(new No565().arrayNesting(nums));
    }
}
